src/Apps/ArtQuiz/ArtQuizContext/ArtQuizContext.tsx

```diff
@@ -0,0 +1,169 @@
+import {
+  createContext,
+  FC,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react"
+import { extractNodes } from "Utils/extractNodes"
+
+// TODO: Kill all the anys!!
+export interface ArtQuizContextValues {
+  artworks: any[]
+  artworksTotalCount: number
+  currentArtwork: any | null
+  currentIndex: number
+  markItemComplete: (
+    artworkSlug: keyof ArtQuizContextValues["quizTemplate"]
+  ) => void
+  nextArtwork: any | null
+  previousArtwork: any | null
+  stepForward: () => void
+  stepBackward: () => void
+  quizTemplate: Record<string, boolean>
+}
+
+export const artQuizContext = createContext<ArtQuizContextValues>({
+  artworks: [],
+  artworksTotalCount: 0,
+  currentArtwork: null,
+  currentIndex: 0,
+  nextArtwork: null,
+  previousArtwork: null,
+  markItemComplete: (_: keyof ArtQuizContextValues["quizTemplate"]) => {
+    throw new Error("ArtQuizContext not initialized")
+  },
+  stepForward: () => {
+    throw new Error("ArtQuizContext not initialized")
+  },
+  stepBackward: () => {
+    throw new Error("ArtQuizContext not initialized")
+  },
+  quizTemplate: {},
+})
+
+export const ArtQuizContextProvider: FC = ({ children }) => {
+  // Quiz template should be the source of truth for the content and state of a
+  // particular user's quiz. A stringified version will be stored in a table
+  // with the userId as its key.
+  const [quizTemplate, setQuizTemplate] = useState<Record<string, boolean>>(
+    (undefined as unknown) as Record<string, boolean>
+  )
+
+  // This typing should clarify itself once we create our query and figure
+  // out which values we need to appropriately render the UI
+  const [artworks, _setArtworks] = useState<any[]>(extractNodes({ edges }))
+  const [currentIndex, setCurrentIndex] = useState(0)
+
+  const currentArtwork = useCallback(() => {
+    if (artworks.length > currentIndex) {
+      return artworks[currentIndex]
+    }
+    return null
+  }, [artworks, currentIndex])
+
+  const nextArtwork = useCallback(() => {
+    if (artworks.length > currentIndex + 1) {
+      return artworks[currentIndex + 1]
+    }
+    return null
+  }, [artworks, currentIndex])
+
+  const previousArtwork = useCallback(() => {
+    if (currentIndex > 0) {
+      return artworks[currentIndex - 1]
+    }
+    return null
+  }, [artworks, currentIndex])
+
+  const stepForward = useCallback(() => {
+    if (artworks.length > currentIndex + 1) {
+      setCurrentIndex(currentIndex + 1)
+    }
+  }, [artworks, currentIndex])
+
+  const stepBackward = useCallback(() => {
+    if (currentIndex > 0) {
+      setCurrentIndex(currentIndex - 1)
+    }
+  }, [currentIndex])
+
+  useEffect(() => {
+    // TODO: check db for quizTemplate
+    // if exists, skip creation logic and set
+    // else, create template for user and save to db
+    if (!quizTemplate) {
+      const template = {}
+      for (let artwork of artworks) {
+        Object.assign(template, { [artwork.slug]: false })
+      }
```

Comment:
Stylistic but find this way more confusing than a simple reduce.

```tsx
const template = artworks.reduce((acc, artwork) => ({ ...acc, [artwork.slug]: false }), {})
```

Reply:
Yah with map/filter/reduce no need to ever really mutate variables and use `for` loops

Reply:
yeah I suggested this before last week and its clean, we're gonna add this in